builtin/stash--helper.c

```diff
@@ -0,0 +1,75 @@
+#include "../cache.h"
+#include "../cache-tree.h"
+#include "../diff.h"
+#include "../diffcore.h"
+#include "../gettext.h"
+#include "../git-compat-util.h"
+#include "../revision.h"
+
+static const char builtin_stash__helper_usage[] = {
+	"Usage: git stash--helper --non-patch <i_tree>"
+};
+
+static void diff_callback(struct diff_queue_struct *q,
+					 struct diff_options *options, void *data)
+{
+	int i;
+    fprintf(stderr,"diff_callback\n");
+	for (i = 0; i < q->nr; i++)
+    {
+		struct diff_filepair *p;
+		struct stat st;
+
+        fprintf(stderr,"diff_callback: %d\n", i);
+		p = q->queue[i];
+		add_to_index(&the_index, p->two->path, &st, 0);
+	}
+}
+
+int stash_non_patch(const char *i_tree,	const char *prefix)
+{
+	int result;
+	struct rev_info rev;
+	const char *wt_prefix = NULL;
+	unsigned char sha1[20];
+	const char *me = "git stash--helper";
+
+    fprintf(stderr,"stash_non_patch\n");
+	read_index(&the_index);
+    fprintf(stderr,"read the index\n");
+	init_revisions(&rev, NULL);
+    fprintf(stderr,"initialised rev\n");
+	setup_revisions(0, NULL, &rev, NULL);
+    fprintf(stderr,"setup rev\n");
+	rev.diffopt.output_format |= DIFF_FORMAT_CALLBACK;
+	rev.diffopt.format_callback = diff_callback;
+	rev.diffopt.format_callback_data=&result;
+	//diff_setup_done(&rev.diffopt);
+    fprintf(stderr,"prepared diffopts\n");
```

Comment:
On Linux I get a slightly different result:

````bash
 GIT_TRACE=1 ./t3903-stash.sh --run=                                                                                                                                                                                               1 -d --verbose-only=1
expecting success:
        echo 1 > file &&
        git add file &&
        echo unrelated >other-file &&
        git add other-file &&
        test_tick &&
        git commit -m initial &&
        echo 2 > file &&
        git add file &&
        echo 3 > file &&
        test_tick &&
        git stash &&
        git diff-files --quiet &&
        git diff-index --cached --quiet HEAD

trace: built-in: git 'add' 'file'
trace: built-in: git 'add' 'other-file'
trace: built-in: git 'commit' '-m' 'initial'
[master (root-commit) fe7e97b] initial
 Author: A U Thor <author@example.com>
 2 files changed, 2 insertions(+)
 create mode 100644 file
 create mode 100644 other-file
trace: built-in: git 'add' 'file'
trace: exec: 'git-stash'
trace: run_command: 'git-stash'
trace: built-in: git 'rev-parse' '--git-dir'
trace: built-in: git 'rev-parse' '--git-path' 'objects'
trace: built-in: git 'rev-parse' '--is-inside-work-tree'
trace: built-in: git 'rev-parse' '--show-toplevel'
trace: built-in: git 'rev-parse' '--git-path' 'index'
trace: built-in: git 'config' '--get-colorbool' 'color.interactive'
trace: built-in: git 'update-index' '-q' '--refresh'
trace: built-in: git 'diff-index' '--quiet' '--cached' 'HEAD' '--ignore-submodul                                                                                                                                                                                               es' '--'
trace: built-in: git 'reflog' 'exists' 'refs/stash'
trace: built-in: git 'rev-parse' '--verify' '--quiet' 'refs/stash'
trace: built-in: git 'update-index' '-q' '--refresh'
trace: built-in: git 'diff-index' '--quiet' '--cached' 'HEAD' '--ignore-submodul                                                                                                                                                                                               es' '--'
trace: built-in: git 'rev-parse' '--verify' 'HEAD'
trace: built-in: git 'rev-list' '--oneline' '-n' '1' 'HEAD' '--'
trace: built-in: git 'symbolic-ref' '-q' 'HEAD'
trace: built-in: git 'write-tree'
trace: built-in: git 'commit-tree' 'bb8058573351acfa9770d8e156931a8a44bf1a9e' '-                                                                                                                                                                                               p' 'fe7e97bbec155b40a99ae689b9aec39d0b0d8b03'
trace: built-in: git 'stash--helper' '--non-patch' 'bb8058573351acfa9770d8e15693                                                                                                                                                                                               1a8a44bf1a9e'
stash_non_patch
read the index
initialised rev
setup rev
prepared diffopts
Segmentation fault (core dumped)
Cannot save the current worktree state
not ok 1 - stash some dirty working directory
#
#               echo 1 > file &&
#               git add file &&
#               echo unrelated >other-file &&
#               git add other-file &&
#               test_tick &&
#               git commit -m initial &&
#               echo 2 > file &&
#               git add file &&
#               echo 3 > file &&
#               test_tick &&
#               git stash &&
#               git diff-files --quiet &&
#               git diff-index --cached --quiet HEAD
#

ok 2 # skip parents of stash (--run)
ok 3 # skip applying bogus stash does nothing (--run)
ok 4 # skip apply does not need clean working directory (--run)
ok 5 # skip apply does not clobber working directory changes (--run)
ok 6 # skip apply stashed changes (--run)
ok 7 # skip apply stashed changes (including index) (--run)
ok 8 # skip unstashing in a subdirectory (--run)
ok 9 # skip stash drop complains of extra options (--run)
ok 10 # skip drop top stash (--run)
ok 11 # skip drop middle stash (--run)
ok 12 # skip stash pop (--run)
ok 13 # skip stash branch (--run)
ok 14 # skip apply -q is quiet (--run)
ok 15 # skip save -q is quiet (--run)
ok 16 # skip pop -q is quiet (--run)
ok 17 # skip pop -q --index works and is quiet (--run)
ok 18 # skip drop -q is quiet (--run)
ok 19 # skip stash -k (--run)
ok 20 # skip stash --no-keep-index (--run)
ok 21 # skip stash --invalid-option (--run)
ok 22 # skip stash an added file (--run)
ok 23 # skip stash rm then recreate (--run)
ok 24 # skip stash rm and ignore (--run)
ok 25 # skip stash rm and ignore (stage .gitignore) (--run)
ok 26 # skip stash file to symlink (missing SYMLINKS)
ok 27 # skip stash file to symlink (stage rm) (missing SYMLINKS)
ok 28 # skip stash file to symlink (full stage) (missing SYMLINKS)
ok 29 # skip stash symlink to file (--run)
ok 30 # skip this must have re-created the symlink (missing SYMLINKS)
ok 31 # skip unstash must re-create the file (--run)
ok 32 # skip stash symlink to file (stage rm) (--run)
ok 33 # skip this must have re-created the symlink (missing SYMLINKS)
ok 34 # skip unstash must re-create the file (--run)
ok 35 # skip stash symlink to file (full stage) (--run)
ok 36 # skip this must have re-created the symlink (missing SYMLINKS)
ok 37 # skip unstash must re-create the file (--run)
ok 38 # skip stash directory to file (--run)
ok 39 # skip stash file to directory (--run)
ok 40 # skip stash branch - no stashes on stack, stash-like argument (--run)
ok 41 # skip stash branch - stashes on stack, stash-like argument (--run)
ok 42 # skip stash show format defaults to --stat (--run)
ok 43 # skip stash show - stashes on stack, stash-like argument (--run)
ok 44 # skip stash show -p - stashes on stack, stash-like argument (--run)
ok 45 # skip stash show - no stashes on stack, stash-like argument (--run)
ok 46 # skip stash show -p - no stashes on stack, stash-like argument (--run)
ok 47 # skip stash drop - fail early if specified stash is not a stash reference                                                                                                                                                                                                (--run)
ok 48 # skip stash pop - fail early if specified stash is not a stash reference                                                                                                                                                                                                (--run)
ok 49 # skip ref with non-existent reflog (--run)
ok 50 # skip invalid ref of the form stash@{n}, n >= N (--run)
ok 51 # skip stash branch should not drop the stash if the branch exists (--run)
ok 52 # skip stash apply shows status same as git status (relative to current di                                                                                                                                                                                               rectory) (--run)
ok 53 # skip stash where working directory contains "HEAD" file (--run)
ok 54 # skip store called with invalid commit (--run)
ok 55 # skip store updates stash ref and reflog (--run)
ok 56 # skip handle stash specification with spaces (--run)
ok 57 # skip setup stash with index and worktree changes (--run)
ok 58 # skip stash list implies --first-parent -m (--run)
ok 59 # skip stash list --cc shows combined diff (--run)
# failed 1 among 59 test(s)
1..59
````

>prepared diffopts
Segmentation fault (core dumped)
Cannot save the current worktree state

That's the segmentation fault I'm talking about. I assume the segmentation fault also happens on windows and just doesn't get a command line message.